Invalid model name results in segmentation violation

When running `gorse serve -c config/gorse.toml` the command crashes:
```
2020/02/05 10:58:07 training model
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x917733]

goroutine 5 [running]:
github.com/zhenghaoz/gorse/engine.UpdateRecommends(0xc000104050, 0xc, 0xc004565110, 0x32, 0x1, 0xb7cdc0, 0xc0000de000, 0xc0000105a0, 0x0, 0x2)
        /root/gorse/engine/offline.go:89 +0x193
github.com/zhenghaoz/gorse/engine.Update(0xc00002bfc0, 0x7, 0x1f90, 0xc000104010, 0xd, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /root/gorse/engine/offline.go:135 +0x1d8
github.com/zhenghaoz/gorse/cmd.watch(0xc00002bfc0, 0x7, 0x1f90, 0xc000104010, 0xd, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /root/gorse/cmd/serve.go:85 +0x36b
created by github.com/zhenghaoz/gorse/cmd.glob..func7
        /root/gorse/cmd/serve.go:31 +0x30c
```

After some investigation I found out that I used an invalid model (`knn-implicit` instead of `knn_implicit`).

IMHO an invalid model should be reported instead of causing a segmentation fault.

--- a/engine/offline.go
+++ b/engine/offline.go
@@ -84,6 +84,10 @@ func UpdateRecommends(name string, params base.Params, cacheSize int, fitJobs in
 	// Create model
 	log.Printf("create model %v with params = %v\n", name, params)
 	model := LoadModel(name, params)
+	if model == nil {
+		log.Fatalf("invalid model %v, aborting", name)
+		return nil
+	}
 	// Training model
 	log.Println("training model")
 	model.Fit(dataSet, &base.RuntimeOptions{Verbose: true, FitJobs: fitJobs})

--- a/engine/offline.go
+++ b/engine/offline.go
@@ -4,7 +4,7 @@ import (
 	"github.com/BurntSushi/toml"
 	"github.com/zhenghaoz/gorse/base"
 	"github.com/zhenghaoz/gorse/core"
-	"errors"
+	"fmt"
 	"log"
 )
 
@@ -86,8 +86,8 @@ func UpdateRecommends(name string, params base.Params, cacheSize int, fitJobs in
 	log.Printf("create model %v with params = %v\n", name, params)
 	model := LoadModel(name, params)
 	if model == nil {
-		log.Fatalf("invalid model %v, aborting", name)
-		return errors.New("invalid model %v")
+		log.Printf("invalid model %v, aborting\n", name)
+		return fmt.Errorf("invalid model %v", name)
 	}
 	// Training model
 	log.Println("training model")

--- a/engine/offline.go
+++ b/engine/offline.go
@@ -4,6 +4,7 @@ import (
 	"github.com/BurntSushi/toml"
 	"github.com/zhenghaoz/gorse/base"
 	"github.com/zhenghaoz/gorse/core"
+	"errors"
 	"log"
 )
 
@@ -86,7 +87,7 @@ func UpdateRecommends(name string, params base.Params, cacheSize int, fitJobs in
 	model := LoadModel(name, params)
 	if model == nil {
 		log.Fatalf("invalid model %v, aborting", name)
-		return nil
+		return errors.New("invalid model %v")
 	}
 	// Training model
 	log.Println("training model")

--- a/engine/offline_test.go
+++ b/engine/offline_test.go
@@ -213,3 +213,32 @@ func TestUpdateRecommends(t *testing.T) {
 		t.Fatal(err)
 	}
 }
+
+func TestUpdateRecommendsInvalidModel(t *testing.T) {
+	// Create database
+	fileName := path.Join(core.TempDir, randstr.String(16))
+	db, err := Open(fileName)
+	if err != nil {
+		t.Fatal(err)
+	}
+	// Update recommends
+	dataSet := core.LoadDataFromBuiltIn("ml-100k")
+	itemId := make([]string, dataSet.ItemCount())
+	for itemIndex := 0; itemIndex < dataSet.ItemCount(); itemIndex++ {
+		itemId[itemIndex] = dataSet.ItemIndexer().ToID(itemIndex)
+	}
+	if err = db.InsertItems(itemId, nil); err != nil {
+		t.Fatal(err)
+	}
+	if err = UpdateRecommends("invalid-model", nil, 10, runtime.NumCPU(), dataSet, db); err == nil {
+		t.Fatal("function should return an error")
+	}
+	// Close database
+	if err = db.Close(); err != nil {
+		t.Fatal(err)
+	}
+	// Clean database
+	if err = os.Remove(fileName); err != nil {
+		t.Fatal(err)
+	}
+}

